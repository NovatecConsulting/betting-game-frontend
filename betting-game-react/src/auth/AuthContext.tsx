import React from "react";
import {
    User as OidcUser,
    UserManager,
    UserManagerSettings,
} from "oidc-client";

// For immutability
class User extends OidcUser {
    constructor(user: OidcUser) {
        super({
            id_token: user.id_token,
            access_token: user.access_token,
            expires_at: user.expires_at,
            profile: user.profile,
            refresh_token: user.refresh_token!,
            scope: user.scope,
            session_state: user.session_state!,
            state: user.state,
            token_type: user.token_type,
        });
    }
}

// TODO: error handling
/**
 * Manages the Authorization Code Flow with PKCE. Added such that underlying oidc library can be exchanged very easy
 * without being forced to change other parts of the app.
 */
class AuthorizationCodeFlowPKCEManager {
    private userManager: UserManager;
    private user: User | null;

    constructor(settings: UserManagerSettings) {
        this.userManager = new UserManager(settings);
        this.user = null;
    }

    public login = async () => {
        await this.userManager.clearStaleState();
        await this.userManager.signinRedirect();
    };

    public loginCallback = async () => {
        await this.userManager.signinRedirectCallback();
        const oidcUser = await this.userManager.getUser();
        this.user = oidcUser ? new User(oidcUser) : oidcUser;
    };

    public logout = async () => {
        // this causes an error at the moment, because there is a CORS bug at auth0
        // if a user is logged out the corresponding session and all refresh_tokens etc are revoked. on localhost this
        // has the effect, that users have to consent the access to their email on each login.
        // This is not happening on a real domain.
        // FIXME: add again
        // await this.userManager.revokeAccessToken();
        await this.userManager.signoutRedirect();
    };

    public updateToken = async (minValidity: number = 5) => {
        if (
            this.user &&
            this.user.refresh_token &&
            (this.user.expired || this.user.expires_in === minValidity)
        ) {
            try {
                await this.userManager.signinSilent();
                const oidcUser = await this.userManager.getUser();
                this.user = oidcUser ? new User(oidcUser) : oidcUser;
            } catch (e) {
                console.error(
                    "User context could not be refreshed. Logging out.",
                    e
                );
                await this.logout();
            }
        }
    };

    public getUser = () => this.user;
}

export let AuthorizationManager: AuthorizationCodeFlowPKCEManager;

const AuthContext: React.FC = ({ children }) => {
    // TODO: use config
    AuthorizationManager = new AuthorizationCodeFlowPKCEManager({
        authority: "https://betting-game.eu.auth0.com",
        client_id: "rqlyAphr2es3RUjYlPAEkf2VCldAR5wL",
        redirect_uri: "http://localhost:3000/login-callback",
        response_type: "code",
        response_mode: "query", // TODO: in the future hopefully "fragment" is supported
        scope: "openid profile email offline_access",
        post_logout_redirect_uri: "http://localhost:3000/",
        revokeAccessTokenOnSignout: true,
        metadata: {
            issuer: "https://betting-game.eu.auth0.com/",
            authorization_endpoint:
                "https://betting-game.eu.auth0.com/authorize",
            userinfo_endpoint: "https://betting-game.eu.auth0.com/userinfo",
            token_endpoint: "https://betting-game.eu.auth0.com/oauth/token",
            registration_endpoint:
                "https://betting-game.eu.auth0.com/oidc/register",
            revocation_endpoint:
                "https://betting-game.eu.auth0.com/oauth/revoke",
            end_session_endpoint:
                "https://betting-game.eu.auth0.com/v2/logout?client_id=rqlyAphr2es3RUjYlPAEkf2VCldAR5wL&returnTo=http://localhost:3000/",
            jwks_uri: "https://betting-game.eu.auth0.com/",
        },
    });
    return <>{children}</>;
};

export default AuthContext;
