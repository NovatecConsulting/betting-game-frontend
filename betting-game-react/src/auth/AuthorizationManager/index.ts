export { default as AuthorizationCodeFlowPKCEManager } from "./AuthorizationManager";
