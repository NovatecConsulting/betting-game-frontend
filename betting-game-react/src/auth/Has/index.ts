export { default as Has } from "./Has";
