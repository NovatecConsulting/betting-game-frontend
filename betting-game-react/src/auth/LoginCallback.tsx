import React, { useState } from "react";
import { Redirect } from "react-router-dom";
import { AuthorizationManager } from ".";
import { ROUTE_INDEX } from "../routes";

const LoginCallback: React.FC = () => {
    const [loggedIn, setLoggedIn] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);
    React.useEffect(() => {
        const doLoginAndRedirect = async () => {
            try {
                await AuthorizationManager.loginCallback();
                setLoggedIn(true);
            } catch (e) {
                setError(e.message);
            }
        };
        doLoginAndRedirect();
    }, []);
    // TODO: use Loading Component
    return error ? (
        <div>{error}</div>
    ) : loggedIn ? (
        <Redirect to={ROUTE_INDEX} />
    ) : (
        <div>Loading...</div>
    );
};

export default LoginCallback;
