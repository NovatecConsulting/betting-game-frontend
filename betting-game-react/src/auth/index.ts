export { default as AuthContext, AuthorizationManager } from "./AuthContext";
export { default as LoginCallback } from "./LoginCallback";
export * from "./permissions";
