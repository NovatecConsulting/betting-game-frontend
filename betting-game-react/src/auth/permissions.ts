export const READ_TEAMS = "read:teams";
export const READ_MATCHDAYS = "read:matchdays";
