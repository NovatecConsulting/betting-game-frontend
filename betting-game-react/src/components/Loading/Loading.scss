.loader {
    border-top-color: #1e3a8a;
    -webkit-animation: spinner 1s linear infinite;
    animation: spinner 1s linear infinite;
}

@-webkit-keyframes spinner {
    0% {
        -webkit-transform: rotate(0deg);
    }
    100% {
        -webkit-transform: rotate(360deg);
    }
}

@keyframes spinner {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
}
