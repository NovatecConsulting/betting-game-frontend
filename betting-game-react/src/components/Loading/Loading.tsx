import React from "react";
import "./Loading.scss";

interface Props {
    size?: number;
}

const Loading: React.FC<Props> = ({ size = 16 }) => {
    const style = `loader ease-linear rounded-full border-4 border-t-4 border-gray-200 h-${size} w-${size}`;
    return (
        <div className="flex justify-center mt-8">
            <div className={style}></div>
            <span className="sr-only">Loading...</span>
        </div>
    );
};

export default Loading;
