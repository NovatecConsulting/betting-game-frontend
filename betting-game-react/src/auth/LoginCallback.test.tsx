import React from "react";
import { shallow } from "enzyme";
import LoginCallback from "./LoginCallback";
import { AuthorizationManager } from "../auth";
import { Redirect } from "react-router";
import { Loading } from "../components/Loading";

jest.mock("../auth", () => ({
    AuthorizationManager: {
        loginCallback: jest.fn(),
    },
}));

describe(LoginCallback.name + " component", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    it("should render and match snapshot with loading component before anything happens.", async () => {
        const setLoggedInMock = jest.fn();
        const useStateMock: any = () => [false, setLoggedInMock];
        const errorMock = jest.fn();
        const useStateMock2: any = () => [undefined, errorMock];

        jest.spyOn(React, "useState")
            .mockImplementationOnce(useStateMock)
            .mockImplementationOnce(useStateMock2);

        const wrapper = await shallow(<LoginCallback />);

        expect(AuthorizationManager.loginCallback).toBeCalledTimes(0);
        expect(setLoggedInMock).toHaveBeenCalledTimes(0);
        expect(errorMock).toHaveBeenCalledTimes(0);
        expect(wrapper.find(Loading)).toExist();
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot with successful redirect.", async () => {
        const setLoggedInMock = jest.fn();
        const useStateMock: any = () => [true, setLoggedInMock];
        const errorMock = jest.fn();
        const useStateMock2: any = () => [undefined, errorMock];

        jest.spyOn(React, "useState")
            .mockImplementationOnce(useStateMock)
            .mockImplementationOnce(useStateMock2);
        jest.spyOn(React, "useEffect").mockImplementationOnce((f) => f());
        (AuthorizationManager.loginCallback as jest.Mock).mockImplementationOnce(
            () => Promise.resolve()
        );

        const wrapper = await shallow(<LoginCallback />);

        expect(AuthorizationManager.loginCallback).toBeCalledTimes(1);
        expect(setLoggedInMock).toHaveBeenCalledTimes(1);
        expect(setLoggedInMock).toHaveBeenCalledWith(true);
        expect(errorMock).toHaveBeenCalledTimes(0);
        expect(wrapper.find(Redirect)).toExist();
        expect(wrapper.find(Redirect).prop("to")).toEqual("/");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if AuthorizationManager throws an error.", async () => {
        const setLoggedInMock = jest.fn();
        const useStateMock: any = () => [false, setLoggedInMock];
        const errorMock = jest.fn();
        const useStateMock2: any = () => ["TestError", errorMock];
        const mockPromise = Promise.reject({ message: "TestError" });

        jest.spyOn(React, "useState")
            .mockImplementationOnce(useStateMock)
            .mockImplementationOnce(useStateMock2);
        jest.spyOn(React, "useEffect").mockImplementationOnce((f) => f());
        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.loginCallback as jest.Mock).mockReturnValueOnce(
            mockPromise
        );

        const wrapper = await shallow(<LoginCallback />);

        await mockPromise.catch(() => {}); // Wait for promise

        expect(AuthorizationManager.loginCallback).toBeCalledTimes(1);
        expect(setLoggedInMock).toHaveBeenCalledTimes(0);
        expect(errorMock).toHaveBeenCalledTimes(1);
        expect(errorMock).toHaveBeenCalledWith("TestError");
        expect(wrapper).toIncludeText("TestError");
        expect(wrapper).toMatchSnapshot();
    });
});
