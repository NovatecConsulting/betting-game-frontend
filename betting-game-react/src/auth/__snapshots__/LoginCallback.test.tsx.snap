// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginCallback component should render and match snapshot with loading component before anything happens. 1`] = `<Loading />`;

exports[`LoginCallback component should render and match snapshot with successful redirect. 1`] = `
<Redirect
  to="/"
/>
`;

exports[`LoginCallback component should render and match snapshot, if AuthorizationManager throws an error. 1`] = `
<div>
  TestError
</div>
`;
